src/main/java/com/commercetools/project/sync/SyncerFactory.java

```diff
@@ -55,122 +61,172 @@ public static SyncerFactory of(
-            ? extends BaseSync<?, ?, ?>>
-        syncer;
+  private static Collection<List<SyncModuleOption>> groupSyncModuleOptions(
+      List<SyncModuleOption> syncModuleOptions) {
+    Map<SyncModuleOption, Integer> syncModuleOptionWithReferencesCount =
+        syncModuleOptions
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    key -> key,
+                    option ->
+                        (option.getReferences().isEmpty())
+                            ? 0
+                            : countReferences(option.getReferences(), syncModuleOptions)));
 
-    try {
-      syncer = buildSyncer(syncOptionValue, runnerNameOptionValue, isSyncProjectSyncCustomObjects);
-    } catch (IllegalArgumentException exception) {
+    Map<Integer, List<SyncModuleOption>> groupedSyncModuleOptions = new TreeMap<>();
+    for (Map.Entry<SyncModuleOption, Integer> syncOptionWithReferenceCount :
+        syncModuleOptionWithReferencesCount.entrySet()) {
+      if (!groupedSyncModuleOptions.containsKey(syncOptionWithReferenceCount.getValue())) {
+        List<SyncModuleOption> syncOptionsList = new ArrayList<>();
+        for (Map.Entry<SyncModuleOption, Integer> syncModuleOptionIntegerEntry :
+            syncModuleOptionWithReferencesCount.entrySet()) {
+          if (0
+              == syncOptionWithReferenceCount
+                  .getValue()
+                  .compareTo(syncModuleOptionIntegerEntry.getValue())) {
+            syncOptionsList.add(syncModuleOptionIntegerEntry.getKey());
+          }
+        }
+        groupedSyncModuleOptions.put(syncOptionWithReferenceCount.getValue(), syncOptionsList);
+      }
+    }
+    return groupedSyncModuleOptions.values();
+  }
 
-      return exceptionallyCompletedFuture(exception);
+  private static int countReferences(
+      List<SyncModuleOption> syncOptionDependencies, List<SyncModuleOption> syncModuleOptions) {
+    if (null == syncOptionDependencies || syncOptionDependencies.isEmpty()) {
+      return 1;
+    }
+    int count = 0;
+    for (SyncModuleOption value : syncOptionDependencies) {
+      if (!syncModuleOptions.contains(value)) {
+        continue;
+      }
+      count += countReferences(value.getReferences(), syncModuleOptions);
     }
+    return count;
+  }
 
-    return syncer
-        .sync(runnerNameOptionValue, isFullSync)
-        .whenComplete((syncResult, throwable) -> closeClients());
+  private void closeClients() {
+    sourceClientSupplier.get().close();
+    targetClientSupplier.get().close();
   }
 
   /**
    * Builds an instance of {@link Syncer} corresponding to the passed option value.
    *
-   * @param syncOptionValue the string value passed to the sync option.
+   * @param syncModuleOption the string value passed to the sync option.
    * @return The instance of the syncer corresponding to the passed option value.
    * @throws IllegalArgumentException if a wrong option value is passed to the sync option.
```

Comment:
remove this line. It isn't valid anymore

Reply:
https://github.com/commercetools/commercetools-project-sync/pull/219/commits/938c60a93adc56d2fc60fd9fe83fff6b6a3f4494